### PR TITLE
MEN-1741: modify any file using the mender-artifact tool

### PR DESCRIPTION
**Pull request by oleorhagen**
Changelog: Title

Signed-off-by: Ole Petter <ole.orhagen@cfengine.com>

### CODE REVIEW

File: cli/mender-artifact/main.go [hunk 1]
```diff
@@ -218,12 +220,153 @@ func run() error {
...
 		validate,
 		sign,
 		modify,
+		copy,
+		cat,
 	}
 	return app.Run(os.Args)
 }
+
+func Cat(c *cli.Context) (err error) {
```
Comment (pasinskim) [inline cli/mender-artifact/main.go]: I would move those to separate files instead of leaving in main.go.

File: cli/mender-artifact/debugfs.go [hunk 2]
```diff
@@ -33,10 +34,26 @@ func debugfsCopyFile(file, image string) (string, error) {
...
+	data, err := ioutil.ReadAll(ep)
+	if err != nil {
+		return "", errors.Wrap(err, "Failed to read from stderr-pipe")
+	}
+
+	if len(data) > 0 && strings.Contains(string(data), "File not found") {
+		return "", fmt.Errorf("file %s not found in image", file)
+	}
+	if err = cmd.Wait(); err != nil {
+		os.RemoveAll(tmpDir)
```
Comment (pasinskim) [inline cli/mender-artifact/debugfs.go]: Nice you added this, but what with other failed scenarios? We should probably remove the temp dir for all the error cases.

File: cli/mender-artifact/debugfs.go [hunk 3]
```diff
@@ -33,10 +34,26 @@ func debugfsCopyFile(file, image string) (string, error) {
 	dumpCmd := fmt.Sprintf("dump %s %s", file,
 		filepath.Join(tmpDir, filepath.Base(file)))
 	cmd := exec.Command("debugfs", "-R", dumpCmd, image)
-	if err = cmd.Run(); err != nil {
+	ep, err := cmd.StderrPipe()
```
Comment (pasinskim) [inline cli/mender-artifact/debugfs.go:37]: +1

File: cli/mender-artifact/main.go [hunk 4]
```diff
@@ -218,12 +220,153 @@ func run() error {
...
+		Description: "Copies a file into or out of a mender artifact, or sdimg",
+		UsageText:   "Copy from or into an artifact, or sdimg where either the <src> or <dst> has to be of the form [artifact|sdimg]:<filepath>",
+		Action:      CopyGen,
+	}
+
+	cat := cli.Command{
+		Name:        "cat",
+		Usage:       "cat [artifact|sdimg]:<filesrc> <stdout>",
+		Description: "cat <src> <stdout>",
+		UsageText:   "cats a file from an image or mender-artifact to stdout, <src> has to be of the form [artifact|sdimg]:<filepath>",
```
Comment (pasinskim) [inline cli/mender-artifact/main.go]: For consistency, you should start with capital letter.

File: cli/mender-artifact/main.go [hunk 5]
```diff
@@ -218,12 +220,153 @@ func run() error {
...
+		return cli.NewExitError(fmt.Sprintf("failed to open the partition reader: err: %v", err), 1)
+	}
+	var w io.WriteCloser = os.Stdout
+	if _, err = io.Copy(w, r); err != nil {
+		return cli.NewExitError(fmt.Sprintf("failed to copy from: %s to stdout: err: %v", c.Args().First(), err), 1)
+	}
+	return r.Close()
+}
+
+func CopyGen(c *cli.Context) (err error) {
```
Comment (pasinskim) [inline cli/mender-artifact/main.go]: This one should be moved to a separate file as well.

**Review by pasinskim — CHANGES_REQUESTED** [review]
Looks nice overall. I would at least remove the logic from main.go to separate files. Otherwise, there are some smaller changes needed. 
I can merge this later on and continue refactoring. Should have a PR soon as I don't want to spend whole lot of time on refactoring part.

**Comment by oleorhagen** [comment]
Nice, implemented all your requests now. Btw the refactor looks nice!

**Pull request merged**